File: leetcode/working/133_clone_graph.go
// 133. Clone Graph
// https://leetcode.com/problems/clone-graph/description/
package main

type Node struct {
	Val       int
	Neighbors []*Node
}

func cloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	visited := make(map[int]*Node)

	// DFS: PreOrder
	var clone func(node *Node) *Node
	clone = func(node *Node) *Node {
		if val, exists := visited[node.Val]; exists {
			return val
		}
		newNode := &Node{
			Val:       node.Val,
			Neighbors: make([]*Node, 0, len(node.Neighbors)),
		}

		// Process Current
		visited[node.Val] = newNode

		// Process children
		for _, neighbor := range node.Neighbors {
			newNode.Neighbors = append(newNode.Neighbors, clone(neighbor))
		}
		return newNode
	}
	return clone(node)
}

File: leetcode/graphs/133_clone_graph.py
# 133. Clone Graph
# https://leetcode.com/problems/clone-graph

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node: return None
        '''
        {
            <original_node_1> : <copy_node_1>,
            <original_node_2> : <copy_node_2>
        }
        '''
        original_to_clone_map = {}

        stack = [node]  # original nodes that are to be cloned

        while stack:

            curr = stack.pop()
            original_to_clone_map[node] = Node(curr.val)              # make a copy of root node

            for child in curr.neighbors:
                if child not in original_to_clone_map:
                    stack.append(child)
                    child_copy = Node(child.val)
                    original_to_clone_map[child] = child_copy
                original_to_clone_map[curr].neighbors.append(original_to_clone_map[child])

        
        return original_to_clone_map[node]
